### ISSUE TITLE
Template level Semaphore broker during high load

### ISSUE THREAD

**Issue by reenimathew** [argoproj/argo-workflows#12039]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

**Context:**

1. Single Workflowtemplate having semaphore synchronization (value 200 )
2. Workflow with 4 parallel steps and each step referencing above mentioned workflow template
3. Each workflow takes nearly 8-10 mins to complete
4. Triggered 700 Workflows in parallel 

A total of 2800 (**700 * 4**) locks required to complete all workflows

**Expectation:**

1. All 700 workflows move to running state
2. At any point of time, 200 kubernetes pods should run in parallel
3. All 700 workflows will get completed slowly

**What actually happened?**

1. All 700 workflows went into running state
2. 200 kubernetes pods got created for the first time
3. As and when steps got completed, locks got released, pending workflows got enqueued.
5. But enqueued workflows couldn't acquire lock
6. Kubernetes pod count never went upto 200 after that
7. Post doing workflow controller restart, again 200 kubernetes pods started running

![image](https://github.com/argoproj/argo-workflows/assets/6531836/df4fffb8-c8ed-4a6d-9d3c-40b25f144a18)

**What all did we verify**

1. Added logs to check whether its golang semaphore issue - RULED OUT
2. Added logs to check whether same workflow step is acquiring lock more than once - RULED OUT
3. Added logs to check whether workflow is even able to get into acquire step - ISSUE

**What do we think is happening?**
When there are more workflows in pending queue, and when those get enqueued, many of them run in parallel. Hence peek logic is going for a toss and workflow is missing chance to even acquire the lock. 

**Do we have any solution in mind?**
Adding peek item to the front of the workflow queue
Also add the holderKey (workflow under processing) back to the workflow queue. (This is also not a full proof solution)

Note
** **We tested using v3.4.7**

### Version

v3.4.7

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
We reproduced using EventSource-Sensor combination,

Added Workflow and Workflowtemplate for reference
https://github.com/chaudharirohit2810/Argo-issue-reproducer/tree/main

Variables we changed to reproduce the issue
sleep - 600
activeDeadlineSeconds - to avoid failures
semaphore value - 200
```

### Logs from the workflow controller

```text
*************************************
Logs proving availability of semaphore
*************************************
time="2023-10-19T17:05:54.901Z" level=info msg="argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template acquired by argo-ns/stress-test-workflow-hsrxr/stress-test-workflow-hsrxr-3513372326. Lock availability: 163/200" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template
time="2023-10-19T17:06:05.750Z" level=info msg="argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template acquired by argo-ns/stress-test-workflow-brkgb/stress-test-workflow-brkgb-2335400458. Lock availability: 162/200" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template

*************************************
"Enqueue the workflow" were also present
*************************************


*************************************
Logs around peek of pending priority queue not matching
*************************************

time="2023-10-19T12:17:35.690Z" level=info msg="Adding the peek item to workflow queue Peek=argo-ns/stress-test-workflow-8mkbd/stress-test-workflow-8mkbd-916766346 Holder Key=argo-ns/stress-test-workflow-892hh/stress-test-workflow-892hh-1715903970" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template
time="2023-10-19T12:17:35.690Z" level=info msg="argo-ns/stress-test-workflow-892hh/stress-test-workflow-892hh-1715903970 is not the peek, returning back" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template
time="2023-10-19T12:17:35.741Z" level=info msg="Adding the peek item to workflow queue Peek=argo-ns/stress-test-workflow-8mkbd/stress-test-workflow-8mkbd-916766346 Holder Key=argo-ns/stress-test-workflow-dd2jz/stress-test-workflow-dd2jz-1978979235" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template
time="2023-10-19T12:17:35.741Z" level=info msg="argo-ns/stress-test-workflow-dd2jz/stress-test-workflow-dd2jz-1978979235 is not the peek, returning back" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template
time="2023-10-19T12:17:35.794Z" level=info msg="Adding the peek item to workflow queue Peek=argo-ns/stress-test-workflow-8mkbd/stress-test-workflow-8mkbd-916766346 Holder Key=argo-ns/stress-test-workflow-744nx/stress-test-workflow-744nx-3346821594" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template
time="2023-10-19T12:17:35.794Z" level=info msg="argo-ns/stress-test-workflow-744nx/stress-test-workflow-744nx-3346821594 is not the peek, returning back" semaphore=argo-ns/ConfigMap/dr-stress-test-configmap/dr.test.throttle.template
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

**Comment by reenimathew** [comment]
Fixes we tried

1) Right now we publish the peak workflow in the pending queue back to workflow queue (when workflow name under processing doesn't match the peek ). We tried publishing processed workflow also back to the queue.
2) We tried publishing processed workflow back to the queue with a delay of 20sec
3) We tried publishing only peek instead of publishing first N workflows when we release / acquire semaphore (NextWorkflow function call)

None of the above worked as expected

**Comment by tooptoop4** [comment]
maybe related to https://github.com/argoproj/argo-workflows/issues/11984

**Comment by icecoffee531** [comment]
I have the same issue — the configured semaphore doesn't match the actual number of concurrently running workflows. It seems like the lock is not being released quickly enough to keep up with the execution speed of the tasks.

**Comment by brokenjacobs** [comment]
I'm seeing this issue as well, anything that can be done to help troubleshoot? it's so odd, the ui will show something like 10/200 locks but it won't start anything. This is on argo 4.0.2.

**Comment by brokenjacobs** [comment]
can anyone on this thread confirm this issue doesn't impact workflow level semaphores?

**Comment by reenimathew** [comment]
This doesn't impact workflow level semaphores. It's specific to Workflowtemplate.

**Comment by brokenjacobs** [comment]
@reenimathew would db level semaphores mitigate this for now?

**Comment by brokenjacobs** [comment]
> This doesn't impact workflow level semaphores. It's specific to Workflowtemplate.

I'm confused, is this issue about template level semaphores inside a WorkflowTemplate, but inside a Workflow it works? My semaphore reference is only in the child workflows, not the workflow template instantiating them, but I still see this concurrency issue. Concurrency gets limtied up the semaphore size, as workflows complete no new workflows start because locks are not released. I'm trying to enforce 'pod level' concurrency to keep the cluster busy across multiple workflows. Is this a separate issue from this one?

For example, start 60x workflows from a template, each workflow instantiates 40 dag steps in parallel using withParam for a total of 2400 dag nodes. I use a semaphore to enforce 1000 to run at once and it works, but 1400 don't run.  The lock status will go down to like 2/1000 in the UI but the lock is never acquired, unless I restart the workflow controller.

**Comment by brokenjacobs** [comment]
I can confirm the same issue occurs using a DB semaphore as it does using a controller semaphore. Controller screenshot:

<img width="335" height="96" alt="Image" src="https://github.com/user-attachments/assets/d1d9704d-699e-4446-a648-2ef953461a4b" />

It's not starting more work.

**Comment by yinonv** [comment]
We hit the same throughput collapse and traced it to **write amplification on pending workflows**.

Every *waiting* workflow stores a copy of the global lock state in its own status, so any acquire/release rewrites **all** pending workflows to etcd even though nothing changed for them.

Here's how a pending workflow object looks while waiting on the lock:

```yaml
# pending workflow .status
Message: Waiting for .../my-semaphore/workflow lock. Lock status: 0/1
Synchronization:
  Semaphore:
    Waiting:
      Holders:
        - my-namespace/my-workflow-abc12   # the *other* wf holding the lock
      Semaphore: .../ConfigMap/my-semaphore/workflow
```

Two things make this expensive on **every reconciliation cycle** of a pending workflow:

1. [SemaphoreStatus.LockWaiting()](https://github.com/argoproj/argo-workflows/blob/2fc922d950bc5746001176edce8a095209ee4cee/pkg/apis/workflow/v1alpha1/workflow_types.go#L4120) always returns `true` , unlike the mutex version which compares first. So `wfUpdate=true` every cycle regardless of change.
2. **The waiting message embeds a live counter** (`Lock status: %d/%d` in [semaphore.go](https://github.com/argoproj/argo-workflows/blob/main/workflow/sync/semaphore.go#L209)). For example, with a semaphore of 200, [markWorkflowPhase](https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/operator.go#L2597) sees it change (e.g. `120/200` → `119/200`) whenever *any* workflow acquires/releases, sets `woc.updated=true`, and triggers a full persist.

> The `Holders` list on each waiter is also purely informational. The source of truth is the controller's in-memory `lockHolder` map.

**Combined effect:** every pending workflow gets dehydrated and written to the API server every cycle for zero state change. With hundreds/thousands pending, the controller becomes I/O-bound on pointless writes and can barely promote new workflows.

**Comment by brokenjacobs** [comment]
For anyone stuck like me here trying to scale your workflows up to a large limit using semaphores, what has worked for us it do deploy kueue:
https://kueue.sigs.k8s.io/

It has a semaphore like primitive that works with Argo workflow. (And works)
